### PR TITLE
Replace lots of hamcrest's assert_that with a builder style

### PR DESCRIPTION
**Pull request by dwijnand**
Refs #5742

I've been thinking of possible solutions to close out #5742, so I'm submitting this for early feedback.

@alexcrichton what do you think? If you like the look of this I'll apply it across the suite.

r? @alexcrichton

### CODE REVIEW

File: tests/testsuite/support/mod.rs [hunk 1]
```diff
@@ -990,6 +1003,15 @@ impl Execs {
     }
 }
 
+impl Drop for Execs {
+    fn drop(&mut self) {
+        if !self.ran {
+            // TODO: Re-enable when everything goes through Execs#run
```
Comment (Eh2406) [inline tests/testsuite/support/mod.rs]: Overall I like the direction. mabe add a `#[must_use]` to `Execs`?

Comment (dwijnand) [reply tests/testsuite/support/mod.rs]: I'm copying Alex's implementation in rustfix: https://github.com/rust-lang-nursery/rustfix/blob/116c7e9f7b666d92c049ab86c1d83e8ced1cead4/cargo-fix/tests/all/main.rs#L323-L329

How do the two techniques compare?

Comment (dwijnand) [reply tests/testsuite/support/mod.rs]: Interesting: `ProjectBuilder` already has the `must_use` attribute applied to it.

Comment (Eh2406) [reply tests/testsuite/support/mod.rs]: Defence in depth, use both. One checks at runtime the other at compile time.

**Comment by alexcrichton** [comment]
Looking good to me! I wonder if we could lift the requirement though of requiring both `execs` and `run`? Having both feels slightly redundant (but I think one is good to keep)

**Comment by dwijnand** [comment]
Yeah we can make `.cargo` return `execs` and then forward the `ProcessBuilder` methods used in the test suite.

How do you feel, instead, about running the assert in the Drop? It would remove the need for to expressly execute `.run()` but maybe it's a bad practice? `.run()` would still be there when you want to force the execution order. I'm torn :smile:

**Comment by alexcrichton** [comment]
Sounds plausible to me! (having `.cargo` return `Execs`)

I'd be more wary of removing the need for `.run()` in the sense that having such significant behavior happen in a destructor can be a bit scary for codegen reasons here and there.

**Comment by alexcrichton** [comment]
Looks great to me!

**Comment by dwijnand** [comment]
This doesn't completely remove hamcrest, but it does a lot.

I'm concerned that grand refactorings will clash with parallel development in PRs, so I'd love to land this PR into master and send follow up PRs to close out #5742.

**Comment by dwijnand** [comment]
Could someone restart AppVeyor please?

**Comment by dwijnand** [comment]
Yeah but both are noisy.

That said my commenting is also noisy so I'm not doing any better..

**Comment by alexcrichton** [comment]
Holy cow, nice! Thanks so much for taking this on!

Let's definitely get this in before it bitrots!

@bors: r+

**Comment by Eh2406** [comment]
And another net negative PR. If you keep up with this the project will just disappear. :-P

**Pull request merged**